### ISSUE TITLE
120$: create minimal Monaco editor

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#938]
[Monaco](https://microsoft.github.io/monaco-editor/) is a browser-based code editing widget. In my experience, Monaco performs better than `<textarea>`, so I think we should use Monaco instead of `<textarea>` in the web demo (#164).

The problem with Monaco is that its bundle is *huge* (several megabytes). I want a small download.

Task: Fork Monaco and strip it down to just the features we need:
* diagnostic squigglies
* undo/redo
* syntax highlighting (maybe)
* auto-indent (maybe)
See this config for a stripped down Monaco: https://github.com/quick-lint/quick-lint-js/issues/164#issuecomment-1185999754

Goal: Under 100 KiB minified uncompressed bundle size.

Constraint: I want to be able to update Monaco from upstream occasionally.

**Comment by Erik-Daniel** [comment]
I will do this job

**Comment by ahmafi** [comment]
Hi, I would like to work on this.

**Comment by strager** [comment]
> I think just stripping down features from `vscode` which is the base for `monaco-editor-core` is not reasonable. Because it will become obsolete

Can you give more detail?

I was imagining that a stripped-down Monaco would mostly be a red diff (deleted lines), mostly deleting imports. Is it more complicated than that?

> Maybe even creating something from scratch similar to `monaco-editor` that integrates with quick-lint-js would be a better approach.

How does this compare to maintaining a stripped-down monaco-editor?

**Comment by ahmafi** [comment]
Yes, it will be a red diff, I didn't calculate exactly how much of the code we can strip down, but roughly looking at the code I don't think we can go beyond 50% optimization in size, without things getting complicated more than a simple red diff. And because of that, I don't think it's a good choice.

I there was any chance that VS Code team could do this internally and do something like tree-shaking based on needed features, that would be dope.

I also think that the already minified versions of the Moncao are good enough a website that are used mostly by developers.

**Comment by strager** [comment]
> I don't think we can go beyond 50% optimization in size, without things getting complicated more than a simple red diff.

Okay, this makes sense.

> I also think that the already minified versions of the Moncao are good enough a website that are used mostly by developers.

I don't think so. Performance matters even for developers. Improving load times gives a positive perception of "quick"ness.

**Comment by ahmafi** [comment]
Have you seen [CodeMirror](https://codemirror.net)? It seems a decent alternative to Monaco, and It's separating its features into extensions, so the bundle size doesn't go up. Exactly what I meant about Monaco, that they don't do this kind of stuff. It seems to have a decent [bundle size](https://codemirror.net/examples/bundle/).

**Comment by strager** [comment]
@ahmafi I took a second look at CodeMirror. I think you're right; this might be the option we choose instead of Monaco.
